crates/core/src/table/state.rs

```diff
@@ -347,13 +350,18 @@ impl EagerSnapshot {
         let expression = Expression::Struct(expressions);
         let table_schema = DataType::try_struct_type(fields)?;
 
-        let input_schema = self.files.schema();
+        let input_schema = self.snapshot().inner.scan_row_parsed_schema_arrow()?;
         let input_schema = Arc::new(input_schema.as_ref().try_into_kernel()?);
-        let actions = self.files.clone();
-
         let evaluator =
             ARROW_HANDLER.new_expression_evaluator(input_schema, expression.into(), table_schema);
-        let result = evaluator.evaluate_arrow(actions)?;
+
+        let results = self
+            .files
+            .iter()
+            .map(|file| evaluator.evaluate_arrow(file.clone()))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let result = concat_batches(results[0].schema_ref(), &results)?;
```

Comment:
we still concatenate the add actions table. in a follow-up we should also move this to a stream and expose that via record batch readers in python.

Opened #3811 to track this.